aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -239,6 +240,159 @@ Tensor wrapped_scalar_tensor(
+    .check_all_same_dtype(false)
+    .build();
+  AT_DISPATCH_INTEGRAL_TYPES(elements.scalar_type(), "isin_sorting", [&]() {
+    iter.for_each([&](char** data, const int64_t* strides, int64_t n) {
+      auto* mask_bytes = data[0];
+      const auto* element_bytes = data[1];
+
+      for (int i = 0; i < n - 1; ++i) {
+        auto* mask_data = reinterpret_cast<bool*>(mask_bytes);
+        const scalar_t element_val = *(reinterpret_cast<const scalar_t*>(element_bytes));
+        const scalar_t next_element_val = *(reinterpret_cast<const scalar_t*>(element_bytes + strides[1]));
+        *mask_data = invert ? (element_val != next_element_val) : (element_val == next_element_val);
+
+        mask_bytes += strides[0];
+        element_bytes += strides[1];
+      }
+
+      // Last duplicate mask item is unconditionally true or false based on invert flag.
+      auto* mask_data = reinterpret_cast<bool*>(mask_bytes);
+      *mask_data = invert;
+    });
+  });
+
+  Tensor mask = at::empty(all_elements.sizes(), TensorOptions(elements.device()).dtype(ScalarType::Bool));
+  mask.index_copy_(0, sorted_order, duplicate_mask);
+
+  if (assume_unique) {
+    out.copy_(mask.slice(0, 0, elements.numel()).view_as(out));
+  } else {
+    out.copy_(at::index(mask, {c10::optional<Tensor>(unique_order)}));
+  }
+}
+
+Tensor& isin_out(const Tensor& elements, const Tensor& test_elements, bool assume_unique, bool invert, Tensor& out) {
+  out.resize_as_(elements);
+  if (elements.numel() == 0) {
+    return out;
+  }
+
+  // Condition taken from numpy's implementation.
+  // See https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/lib/arraysetops.py#L575
+  if (test_elements.numel() < static_cast<int64_t>(
+        10.0f * std::pow(static_cast<double>(elements.numel()), 0.145))) {
+    isin_default(elements, test_elements.ravel(), invert, out);
+  } else {
+    isin_sorting(elements, test_elements, assume_unique, invert, out);
+  }
+
+  return out;
+}
+
+Tensor& isin_out(const Tensor& elements, const c10::Scalar test_element, bool assume_unique, bool invert, Tensor& out) {
+  Tensor test_elements = wrapped_scalar_tensor(test_element, elements.device());
+  at::isin_out(out, elements, test_elements, assume_unique, invert);
+  return out;
+}
+
+Tensor& isin_out(const c10::Scalar element, const Tensor& test_elements, bool assume_unique, bool invert, Tensor& out) {
+  Tensor elements = wrapped_scalar_tensor(element, test_elements.device());
+  at::isin_out(out, elements, test_elements, assume_unique, invert);
+  return out;
+}
+
+Tensor isin(const Tensor& elements, const Tensor& test_elements, bool assume_unique, bool invert) {
+  Tensor out = at::empty({0}, TensorOptions(elements.device()).dtype(ScalarType::Bool));
+  at::isin_out(out, elements, test_elements, assume_unique, invert);
+  return out;
+}
+
+Tensor isin(const Tensor& elements, const c10::Scalar test_element, bool assume_unique, bool invert) {
```

Comment:
Do we want to support the test element being a scalar? What would the difference be between == and isin() if a scalar is given to isin()?

Reply:
This is mainly for numpy compatible I think.

Reply:
When test element is a scalar, the result is equivalent to `==` / `torch.eq` when `invert=False` and equivalent to `!=` / `torch.ne` when `invert=True`. @glaringlee is correct that this is for numpy compatibility, since they support scalar test elements.

Reply:
I understand but we don't need to support every feature NumPy does. We're "compatible" with NumPy, we don't have to be identical to it. We can support scalars here if you like but it might be simpler to focus on tensors.

Reply:
We're not fully identical to numpy as it is since we don't support the (scalar, scalar) case, but I don't see a reason not to handle the fairly trivial (tensor, scalar) case since it just involves calling the (tensor, tensor) with a wrapped scalar.